Snow fall from large-scale condensation

pierluigividale opened the pull request:
I have added snow diagnostic variables and a few rules for deciding whether precipitation is rain or snow.
My code is very likely full of syntax errors and/or naive programming, eg. introducing IF statements that are likely to compromise performance.

If someone can give me pointers on how to do this correctly, I can then try to see whether I see a diagnostic.
After that, I would like to introduce a parametrisation similar to the ice parametrisation, but the initial triggers must be BOTH: accumulated snow and soil temperature.

TODO before merge

- [x] documentation (maths + interface)
- [x] tests (`snow=false` has zero snow, otherwise some snow, and `reevaporation=0` also possible)
- [x] decide what to do with snow in convection (skip for now)
- [x] consistent naming of precipitation, rain, snow in variables and output

pierluigividale committed:
Update diagnostic_variables.jl

Added diagnostic variables for snow

pierluigividale committed:
Update large_scale_condensation.jl

Added rule to compute large scale snow depending on lowest level temperature

pierluigividale committed:
Update large_scale_condensation.jl

Expanded rule for large scale snow

pierluigividale committed:
Update convection.jl

Added rules for convective snow

pierluigividale committed:
Update precipitation.jl

Added diagnostics for snow

milankl commented on src/output/variables/precipitation.jl:
```diff
@@ -92,6 +92,105 @@ Fields are: $(TYPEDFIELDS)"""
+    transform::F = (x) -> 1000x     # [m] to [mm]
+end
+
+function output!(
+    output::NetCDFOutput,
+    variable::AbstractSnowRateOutputVariable,
+    acc_variable::AbstractOutputVariable,
+)
```
This is exactly the same as for rain precipitation no? Then we can just make this reuse that code and we should call things `AbstractRateOutputVariable` for consistency, whether its snow or rain, it doesn't matter.

milankl commented on src/dynamics/diagnostic_variables.jl:
```diff
@@ -343,6 +343,19 @@ $(TYPEDFIELDS)"""
     "Rate of large-scale precipitation [m/s], instantaneous"
     precip_rate_convection::GridVariable2D = zeros(GridVariable2D, grid)
 
+     # FROZEN PRECIPITATION (SNOW)
+    "Accumulated large-scale snow [m]"
+    snow_large_scale::GridVariable2D = zeros(GridVariable2D, grid)
+
+    "Accumulated large-scale precipitation [m]"
+    snow_convection::GridVariable2D = zeros(GridVariable2D, grid)
+
+    "Rate of large-scale precipitation [m/s], instantaneous"
+    snow_rate_large_scale::GridVariable2D = zeros(GridVariable2D, grid)
+
+    "Rate of large-scale precipitation [m/s], instantaneous"
+    snow_rate_convection::GridVariable2D = zeros(GridVariable2D, grid)
+
```
yes this looks right!

milankl commented:
Some general questions: 

- Do we need to distinguish between snow fall/precipitation (units of m, or a rate of m/s) and snow cover on the ground (also units of m, even though that's probably meters of water if melted?). The latter I see as an analogy to a sea ice thickness.
- Given the `snow[ij] = max(min(snow[ij], 1), 0)` line it seems you are thinking of snow cover as an area fraction, but I am not sure that makes sense?
- Would albedo simply be set to snow if snow cover > 0? We wouldn't distinguish between fresh snow and old snow?
- Why would the soil temperature be kept at freezing? For SST, ocean shouldn't be supercooled, but soil temperatures could be below freezing? The problem I see at the moment in `st[1,ij] = max(st[1,ij], temp_freeze)` that you could have soil of -40C without snow, once it snows the soil temperature would jump up to freezing!
- For snow covered soil, would you also want to insulate the land by reducing the land-air fluxes?

milankl commented:
Also to be clear, I would let this pull request just implement the snow precipitation, if that works we can make a new pull request for snow cover that actually influences the land. Also at the moment writing the snow fall back into the gridded fields is missing, this needs to be added here

https://github.com/SpeedyWeather/SpeedyWeather.jl/blob/96357e66538e4247b3e310cba7a2d3b9f45d9239/src/physics/column_variables.jl#L134-L140

milankl commented:
And maybe a more general question, would we want to be able to switch snow on/off? I believe so yes, maybe the easiest to have a flag `snow::Bool = true` in `ImplicitCondensation` and `SimplifiedBettsMiller` that way we could disable snow at any time or disable it only for convection etc.

pierluigividale commented:
Hi

So, the ideal thing to do would be to create a snow depth variable, but I did not want to do that before consulting you: not sure what the computational cost is of creating more and more variables, and more and more output streams.

Then, if we have snow depth, there are quite well established formulas that go from snow depth to area coverage of snow: fraction of a grid box covered by snow.

Albedo on land will depend on the fraction of bare soil, the fraction of vegetation, and the fraction of snow cover. For now, the first two are not dealt with for now, they are constant in time, but if we introduce snow, we must have a transition from 0 to 100% snow cover.

We do not want to go snow / no-snow for albedo: that would introduce flickering heat fluxes, looks horrible in time series of the entire surface energy balance, and is prone to generate numerical instability.

With regards to snow freezing thawing, the right thing to do would be to:
a) change the thermal conductivity of the top layer if there is snow cover
b) compute the latent heat exchange when snow melts and to add that to the surface energy balance, but I am not terribly keen to do that to a 20cm soil. This is why I was thinking of an interval in which soil is kept at freezing, but prescribing that analytically is probably not good value for money.

Yes, for the time being I was thinking of thinking of insulation: some way for just damping the soil fluxes. But maybe the best way is to do it in a realistic way. Just worried about costs.

Cheers,
PL

> On 20 Aug 2025, at 11:06, Milan Klöwer ***@***.***> wrote:
> 
> 
> milankl
>  left a comment 
> (SpeedyWeather/SpeedyWeather.jl#817)
>  <https://github.com/SpeedyWeather/SpeedyWeather.jl/pull/817#issuecomment-3204985385>
> Some general questions:
> 
> Do we need to distinguish between snow fall/precipitation (units of m, or a rate of m/s) and snow cover on the ground (also units of m, even though that's probably meters of water if melted?). The latter I see as an analogy to a sea ice thickness.
> Given the snow[ij] = max(min(snow[ij], 1), 0) line it seems you are thinking of snow cover as an area fraction, but I am not sure that makes sense?
> Would albedo simply be set to snow if snow cover > 0? We wouldn't distinguish between fresh snow and old snow?
> Why would the soil temperature be kept at freezing? For SST, ocean shouldn't be supercooled, but soil temperatures could be below freezing? The problem I see at the moment in st[1,ij] = max(st[1,ij], temp_freeze) that you could have soil of -40C without snow, once it snows the soil temperature would jump up to freezing!
> For snow covered soil, would you also want to insulate the land by reducing the land-air fluxes?
> —
> Reply to this email directly, view it on GitHub <https://github.com/SpeedyWeather/SpeedyWeather.jl/pull/817#issuecomment-3204985385>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AZ3DDVGHXIECHIP5L7BNC433OQ3A7AVCNFSM6AAAAACEIJMIX6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTEMBUHE4DKMZYGU>.
> You are receiving this because you authored the thread.
>

pierluigividale commented:
Hi

Indeed, for now I just wanted to test having some snowfall. No surface snow scheme at this stage.
As you will see from the other message, I have not designed a surface snow scheme yet.

Thanks for the other pointers, the question on what else needs to be modified was primary, let me try and do all those things, then test that we have some snowfall.

Cheers,
PL 

> On 20 Aug 2025, at 11:17, Milan Klöwer ***@***.***> wrote:
> 
> 
> milankl
>  left a comment 
> (SpeedyWeather/SpeedyWeather.jl#817)
>  <https://github.com/SpeedyWeather/SpeedyWeather.jl/pull/817#issuecomment-3205062763>
> Also to be clear, I would let this pull request just implement the snow precipitation, if that works we can make a new pull request for snow cover that actually influences the land. Also at the moment writing the snow fall back into the gridded fields is missing, this needs to be added here
> 
> https://github.com/SpeedyWeather/SpeedyWeather.jl/blob/96357e66538e4247b3e310cba7a2d3b9f45d9239/src/physics/column_variables.jl#L134-L140
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/SpeedyWeather/SpeedyWeather.jl/pull/817#issuecomment-3205062763>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AZ3DDVBD6NAQP4Z6HW3Y2JT3OQ4JXAVCNFSM6AAAAACEIJMIX6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTEMBVGA3DENZWGM>.
> You are receiving this because you authored the thread.
>

pierluigividale commented:
Yes, shall do that too.

Cheers,
PL


> On 20 Aug 2025, at 11:20, Milan Klöwer ***@***.***> wrote:
> 
> 
> milankl
>  left a comment 
> (SpeedyWeather/SpeedyWeather.jl#817)
>  <https://github.com/SpeedyWeather/SpeedyWeather.jl/pull/817#issuecomment-3205078359>
> And maybe a more general question, would we want to be able to switch snow on/off? I believe so yes, maybe the easiest to have a flag snow::Bool = true in ImplicitCondensation and SimplifiedBettsMiller that way we could disable snow at any time or disable it only for convection etc.
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/SpeedyWeather/SpeedyWeather.jl/pull/817#issuecomment-3205078359>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AZ3DDVBWEISKSBCCFV5YWA33OQ4WZAVCNFSM6AAAAACEIJMIX6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTEMBVGA3TQMZVHE>.
> You are receiving this because you authored the thread.
>

pierluigividale committed:
Update large_scale_condensation.jl

Changed test for snow vs precipitation to avoid IF and to deal with levels.

pierluigividale committed:
Update convection.jl

Changed the convective rain and snow decision with no IF statement and scanning all layers.

pierluigividale committed:
Update convection.jl

Attempted to include same code for convective rain and snow

pierluigividale committed:
Update precipitation.jl

Removed duplications in function, abstract variables, modified to single AbstractRateOutputVariable

pierluigividale committed:
Update column_variables.jl

Added diagnostic for all variants of snow variables

pierluigividale committed:
Update precipitation.jl

Eliminated stupid line with "------"

milankl committed:
Move snow/rain decision into vertical loops

milankl committed:
update changelog

milankl reviewed (commented):
Just some comments on what I changed compared to your version!

milankl commented:
> not sure what the computational cost is of creating more and more variables, and more and more output streams.

This is 2D and independent columns, so it scales linearly with the number of horizontal grid cells and is a constant with the vertical resolution, so don't worry about costs and memory. If we're worried about speed we have to invest more into the transforms or maybe write a semi-Lagrangian advection scheme one day.

milankl committed:
unpack freezing_threshold

milankl committed:
rename to correct LargeScaleSnowOutput

pierluigividale commented:
> > not sure what the computational cost is of creating more and more variables, and more and more output streams.
> 
> This is 2D and independent columns, so it scales linearly with the number of horizontal grid cells and is a constant with the vertical resolution, so don't worry about costs and memory. If we're worried about speed we have to invest more into the transforms or maybe write a semi-Lagrangian advection scheme one day.

OK, cool. I went walking for a bit, and I was thinking what to do here. Good to know that I do not need to worry.

milankl committed:
reset snow in column

milankl committed:
cap convection precip at zero

milankl commented:
Let it snow, let it snow, let it snow. But it snows in funny places! 1 Month (March) accumulated after some spinup in Jan+Feb

<img width="916" height="414" alt="image" src="https://github.com/user-attachments/assets/c6e57cda-fe5b-4bd5-96d7-17acae403153" />

<img width="916" height="414" alt="image" src="https://github.com/user-attachments/assets/282f9a4a-bc21-45a1-ad0a-1daab3197a1d" />

milankl commented on src/physics/define_column.jl:
```diff
@@ -91,10 +91,16 @@ $(TYPEDFIELDS)"""
+    precip_rate_convection::NF = 0          # precipitation rate due to convection [m/s]
     precip_rate_large_scale::NF = 0         # precipitation rate due to large-scale condensation [m/s]
+    
+    # same for snow
+    snow_convection::NF = 0                 # snow due to convection [m]
+    snow_large_scale::NF = 0                # snow due to large-scale condensation [m]
+    snow_rate_convection::NF = 0            # snow rate due to convection [m/s]
+    snow_rate_large_scale::NF = 0           # snow rate due to large-scale condensation [m/s]
```
Yes, not IC. We currently use `ColumnVariables` to extract one column before the parameterizations for that column are evaluated, this makes sure we have somewhat contiguous memory access as the parameterizations access things in the vertical which is far apart in memory. You see this logic here

https://github.com/SpeedyWeather/SpeedyWeather.jl/blob/2085fa1badd20e052f1ac584af99ae27ff32698b/src/physics/tendencies.jl#L20-L36

you reset the column memory, get the data from the diagnostic variables read it into the column, evaluate the parameterzations on it then write back into the tendencies/diagnostics and reuse the same column memory for the next grid point.

pierluigividale commented:
Yes, I think that the problem here is that we are scanning the entire column, which of course is going to cause quite a bit of frozen precipitation to be diagnosed. In reality, as you say, when such precipitation starts falling, it will melt.
For now, maybe, we could limit the levels over which we make the simple threshold-based decision?

pierluigividale commented:
Do you really want to S&M, or would it be best to re-think the criterion?

pierluigividale commented:
OK, is this similar to allocate/deallocate in F90?

PL


> On 20 Aug 2025, at 17:35, Milan Klöwer ***@***.***> wrote:
> 
> 
> @milankl commented on this pull request.
> 
> In src/physics/define_column.jl <https://github.com/SpeedyWeather/SpeedyWeather.jl/pull/817#discussion_r2288557957>:
> 
> > +    # same for snow
> +    snow_convection::NF = 0                 # snow due to convection [m]
> +    snow_large_scale::NF = 0                # snow due to large-scale condensation [m]
> +    snow_rate_convection::NF = 0            # snow rate due to convection [m/s]
> +    snow_rate_large_scale::NF = 0           # snow rate due to large-scale condensation [m/s]
> Yes, not IC. We currently use ColumnVariables to extract one column before the parameterizations for that column are evaluated, this makes sure we have somewhat contiguous memory access as the parameterizations access things in the vertical which is far apart in memory. You see this logic here
> 
> https://github.com/SpeedyWeather/SpeedyWeather.jl/blob/2085fa1badd20e052f1ac584af99ae27ff32698b/src/physics/tendencies.jl#L20-L36
> 
> you reset the column memory, get the data from the diagnostic variables read it into the column, evaluate the parameterzations on it then write back into the tendencies/diagnostics and reuse the same column memory for the next grid point.
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/SpeedyWeather/SpeedyWeather.jl/pull/817#discussion_r2288557957>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AZ3DDVHW2AIXH2XE26V6X4D3OSIVHAVCNFSM6AAAAACEIJMIX6VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZTCMZXGE3TOMZWGY>.
> You are receiving this because you authored the thread.
>

milankl commented:
> Do you really want to S&M, or would it be best to re-think the criterion?

I think we should rework how reevaporation + melting + freezing on the way down should work. Like wouldn't it be a better assumption to say snow is created below some freezing threshold but the column to the ground has to be below freezing (0˚C) for snow to reach the ground, otherwise convert it back to precipitation? Relatedly, what happens if convection occurs in a layer with excess humidity. This falls down a rain but if it falls through a layer where the convection would cause drying then does the rain reevaporate?

milankl commented:
[Frierson et al. 2006](https://journals.ametsoc.org/view/journals/atsc/63/10/jas3753.1.xml) do this

<img width="412" height="228" alt="image" src="https://github.com/user-attachments/assets/e8e353cb-11f2-431d-a789-10e354a69c13" />

I don't want this to be too complicated but there might be something simple yet more sophisticated possible

pierluigividale committed:
Update thermodynamics.jl

Added CC constants for ice species

pierluigividale committed:
Update large_scale_condensation.jl

Added the latent heat exchange associated with freezing precipitation into snow.

milankl committed:
Merge branch 'main' into plv/snow

pierluigividale commented on src/physics/large_scale_condensation.jl:
```diff
@@ -113,6 +113,19 @@ function large_scale_condensation!(
+
+                # temperature tendency from freezing
+                δT_freeze = -(Lᵢ / cₚ) * δq_freeze
+                temp_tend[k] += δT_freeze
```
which also requires commenting out the first δT calculation

pierluigividale committed:
Update large_scale_condensation.jl

Simplified the test for enthalpy calculation when snow forms.
Added a test to melt snow if any layers below the layer of snow formation are above a melting threshold. This is quite conservative, in that a single warm layer below k is enough to destroy snow at k.

pierluigividale committed:
Update large_scale_condensation.jl

Changed the logic for melting snow, as before it was not possible to assign enthalpy to the layer that actually melts the snow.

pierluigividale committed:
Update large_scale_condensation.jl

Fixed units and inserted a check for available snow to melt

pierluigividale committed:
Update large_scale_condensation.jl

Forgot to end the IF loop

pierluigividale committed:
Update large_scale_condensation.jl

Many typos and unit inconsistencies

pierluigividale committed:
Update large_scale_condensation.jl

Had forgotten to change snow for snow_flux near the end

pierluigividale committed:
Update large_scale_condensation.jl

There was a conceptual error with where/how snow_flux is used to update the column snow.

pierluigividale committed:
Update convection.jl

Implemented melting of snow into the BM scheme.
There may be some conceptual errors, someone needs to check this.

milankl committed:
Merge branch 'main' into plv/snow

pierluigividale committed:
Update convection.jl

There was a missing definition for unit conversions, moved a statement up before the snow melting block.

pierluigividale committed:
Update convection.jl

Fixed an error with the unit corrections, as well as removed calculation from the loop, as it is not necessary to compute the correction for each layer.

milankl committed:
Reevaporation, sublimation and snow fall

milankl committed:
Merge branch 'main' into plv/snow

milankl commented:
20 days accumulated snow after 20 days of spin up. Only snows in mountaineous areas, check, and at high latitudes (this is NH winter), check. But it also snows a little in Africa which seems to be at night because it cools down quite a lot. Maybe something that will change if run for longer

<img width="917" height="415" alt="image" src="https://github.com/user-attachments/assets/22da69fc-e717-4759-ba0a-cd3385df5ec0" />

milankl commented:
We need to see how adapt the same ideas to convection and make it consistent. Or we only have snow from condensation also fine with me. 

Then we need to update the docs with the new maths and additional options for users. And add some tests using the snow and reevaporation options. Testing snow I would just check that snow is zero if option snow is false. 

And I would like to do a quick performance check how expensive condensation now is compared to convection.

milankl committed:
bugs + comment out snow from convection

milankl committed:
Merge branch 'plv/snow' of https://github.com/SpeedyWeather/SpeedyWeather.jl into plv/snow

milankl committed:
latent heat release was overwritten

milankl committed:
avoid negative precipitation

milankl commented:
Snow fall NH winter at T63L8

https://github.com/user-attachments/assets/1fae1066-5ad0-4890-82c1-f567a61df36a

milankl commented:
And snow fall in Jan year 2, looks stable to me, that's good. More snow in central Asia and North America, looks good to me.
https://github.com/user-attachments/assets/7a4a85c4-39f8-4ec6-aa4f-6a3a84e20c00

What's not clear to me is how we would do snow in convection, it's a little trickier to conserve enthalpy there in the BettsMiller scheme. But maybe it's good enough to start to only have snow from large-scale condensation?

Performance is also okay: Depending on how much there's to condense a L8 column takes 20-100ns and zero allocations so that's all fine

milankl committed:
snow removed from convection

milankl committed:
Don't output snow convection variable

milankl committed:
some bugs and docs

milankl committed:
Merge branch 'main' into plv/snow

milankl commented:
I've started writing some docs but then also realised some bugs in the code and so this pull request stays a draft until this is all sorted, sorry! Feel free to have a look at the docs already though or contribute writing the last bits of the math!

milankl committed:
rain/snow fluxes in m/s

milankl committed:
Merge branch 'plv/snow' of https://github.com/SpeedyWeather/SpeedyWeather.jl into plv/snow

milankl committed:
docs updated

milankl committed:
large-scale condensation docs updated

milankl committed:
tests added

pierluigividale committed:
Update large_scale_condensation.md

Fixed a typo, added a comment where I thought that the idea was unclear.

pierluigividale committed:
Update large_scale_condensation.md

Tried to improve the paragraph about re-evaporation.

milankl committed:
docs updates

pierluigividale commented:
The beauty of parametrisation, we end up with parameters, and no idea what to set them to, because such things do not exist in nature. Set it to 30 for now. I don't think we should try tuning anything until we have the new radiation parametrisation.

I have read through the documentation, and I have made a couple of suggestions. I think that it would be better to say re-evaporation, it is more readable, but then the flag is reevaporation. Not sure we can put a hyphen in a flag name?

milankl committed:
Merge branch 'plv/snow' of https://github.com/SpeedyWeather/SpeedyWeather.jl into plv/snow

milankl committed:
longwave tests corrected

milankl committed:
docs math typo

milankl commented:
Okay all checks pass, we have docs and tests and will make `snow=true` the default with this. Remaining question I have is on terminology. We now use 

- precipitation for rain, not snow because that's what we called it before we had snow
- rain for liquid precipitation but that is only really used internally or in the docs
- snow for frozen precipitation

Should we use `precipitation` now for rain+snow? And rename everything that's called `precip` in the output netCDFs `rain`? But then retain `total_precipitation` for the sum of rain+snow?

pierluigividale commented:
OK, all species that fall from the sky are precipitation, so all sorts of drops, ice crystals etc. 
I would say precipitation = liquid _ frozen (rain + snow + ...)
Here we are very liberal with the thing that we call snow, it is actually frozen precipitation, but for the time being, without a microphysics scheme, that is all we can do.

So:
1) rename precipitation as rain in NetCDF output
2) total_precipitation = large_scale_rain + convective_rain + large_scale_snow + convective_snow (does not matter we do not have that one yet).

pierluigividale commented:
Should I wait to squash and merge? I would do those things now, but I am on family duty.

milankl committed:
precipitation = rain + snow renamings

pierluigividale commented:
Hmmm, I was about to start, I see you have done it already, but checks are failing. Shall wait a little longer for the end.

milankl committed:
consistent naming

milankl committed:
docs typo

pierluigividale commented:
It is finally time to press the button!
